crates/runtime/src/datafusion/param_utils.rs

```diff
@@ -0,0 +1,277 @@
+use std::collections::HashMap;
+
+use datafusion::{common::ParamValues, scalar::ScalarValue};
+use serde_json::Value;
+use snafu::prelude::*;
+
+#[derive(Snafu, Debug)]
+pub enum Error {
+    #[snafu(display("Nested arrays or objects are not supported as parameter values"))]
+    NestedValues,
+    #[snafu(display("Unsupported JSON number format"))]
+    UnsupportedJsonNumberFormat,
+    #[snafu(display("Parameters must be a JSON array or an object"))]
+    JsonArrayOrObjectRequired,
+}
+
+/// Converts a serde_json::Value into a datafusion::common::ParamValues.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ParamValues, String>` - The converted ParamValues or an error message.
+///
+/// # Supported JSON Formats
+/// - Array: Converted to positional parameters (ParamValues::Positional).
+/// - Object: Converted to named parameters (ParamValues::Named).
+/// - Primitive types (null, bool, number, string) are supported within arrays or objects.
+///
+/// # Errors
+/// - Returns an error if the top-level JSON is not an array or object.
+/// - Returns an error if any value cannot be converted to a ScalarValue.
+pub fn convert_json_to_param_values(json: Value) -> Result<ParamValues, Error> {
+    match json {
+        Value::Array(arr) => {
+            let mut vec = Vec::with_capacity(arr.len());
+            for val in arr {
+                let scalar = json_to_scalar(&val)?;
+                vec.push(scalar);
+            }
+            Ok(ParamValues::List(vec))
+        }
+        Value::Object(obj) => {
+            let mut map = HashMap::new();
+            for (key, val) in obj {
+                let scalar = json_to_scalar(&val)?;
+                map.insert(key, scalar);
+            }
+            Ok(ParamValues::Map(map))
+        }
+        _ => Err(Error::JsonArrayOrObjectRequired),
+    }
+}
+
+/// Helper function to convert a single serde_json::Value to a DataFusion ScalarValue.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ScalarValue, String>` - The converted ScalarValue or an error message.
+///
+/// # Supported JSON Types
+/// - Null -> ScalarValue::Utf8(None)
+/// - Bool -> ScalarValue::Boolean(Some(bool))
+/// - Number -> ScalarValue::Int64(Some(i64)) if integer, else ScalarValue::Float64(Some(f64))
+/// - String -> ScalarValue::Utf8(Some(String))
+/// - Arrays and objects return an error (handled at a higher level).
+fn json_to_scalar(json: &Value) -> Result<ScalarValue, Error> {
+    match json {
+        Value::Null => Ok(ScalarValue::Utf8(None)),
+        Value::Bool(b) => Ok(ScalarValue::Boolean(Some(*b))),
+        Value::Number(n) => {
+            if let Some(i) = n.as_i64() {
```

Comment:
should we try `if let Some(u) = n.as_u64()` first?

https://github.com/influxdata/influxdb3_core/blob/main/iox_query_params/src/params.rs#L539

Reply:
I think the difference will be in how positive numbers will be treated - I believe (based on current code) we will parse them always as Int64 and InfluxDB as UInt64? Honestly I don't know the real impact/what should be the right behavior but my assumption is that Influx probably know what they are doing